faunadb-common/src/main/java/com/faunadb/common/Connection.java

```diff
@@ -112,57 +127,80 @@ public Builder withMetrics(MetricRegistry registry) {
-        r = new MetricRegistry();
+        registry = new MetricRegistry();
       else
-        r = metricRegistry;
+        registry = metricRegistry;
 
-      AsyncHttpClient c;
+      AsyncHttpClient httpClient;
       if (client == null) {
-        AsyncHttpClientConfig config = new DefaultAsyncHttpClientConfig.Builder()
-          .setConnectTimeout(DEFAULT_CONNECTION_TIMEOUT_MS)
-          .setRequestTimeout(DEFAULT_REQUEST_TIMEOUT_MS)
-          .setPooledConnectionIdleTimeout(DEFAULT_IDLE_TIMEOUT_MS)
-          .setMaxRequestRetry(0)
-          .build();
-        c = new DefaultAsyncHttpClient(config);
-      } else
-        c = client;
+        httpClient = new DefaultAsyncHttpClient(
+          new DefaultAsyncHttpClientConfig.Builder()
+            .setConnectTimeout(DEFAULT_CONNECTION_TIMEOUT_MS)
+            .setRequestTimeout(DEFAULT_REQUEST_TIMEOUT_MS)
+            .setPooledConnectionIdleTimeout(DEFAULT_IDLE_TIMEOUT_MS)
+            .setMaxRequestRetry(DEFAULT_MAX_RETRIES)
+            .build()
+        );
+      } else {
+        httpClient = client;
+      }
 
       URL root;
       if (faunaRoot == null)
-        root = new URL("https://rest.faunadb.com");
+        root = FAUNA_ROOT;
       else
         root = faunaRoot;
 
-      return new Connection(root, authToken, c, r);
+      return new Connection(root, authToken, httpClient, registry, new AtomicInteger(1));
     }
   }
 
-  private static String XFaunaDBHost = "X-FaunaDB-Host";
-  private static String XFaunaDBBuild = "X-FaunaDB-Build";
+  private static final String ASCII = "ASCII";
+  private static final String X_FAUNADB_HOST = "X-FaunaDB-Host";
+  private static final String X_FAUNADB_BUILD = "X-FaunaDB-Build";
 
   private final URL faunaRoot;
-  private final String authToken;
   private final String authHeader;
   private final AsyncHttpClient client;
   private final MetricRegistry registry;
+  private final AtomicInteger refCount;
 
   private final Logger log = LoggerFactory.getLogger(getClass());
   private final ObjectMapper json = new ObjectMapper();
 
-  Connection(URL faunaRoot, String authToken, AsyncHttpClient client, MetricRegistry registry) throws UnsupportedEncodingException {
+  private Connection(URL faunaRoot, String authToken, AsyncHttpClient client, MetricRegistry registry, AtomicInteger refCount) {
     this.faunaRoot = faunaRoot;
-    this.authToken = authToken;
     this.authHeader = generateAuthHeader(authToken);
     this.client = client;
     this.registry = registry;
+    this.refCount = refCount;
+  }
+
+  /**
+   * Creates a new short live connection sharing its underneath {@link AsyncHttpClient}. Queries submited to a
+   * session connection will be authenticated with the token provided.
+   *
+   * @param authToken token that will be used to authenticate requests
+   * @return a new {@link Connection}
+   */
+  public Connection newSessionConnection(String authToken) {
+    if (refCount.incrementAndGet() > 1)
```

Comment:
I see you point now, thanks!

The problem I see is that, even if you never use the connection, you still hold a valid referent to it. If we don't increase the `refCount` when we create new connections, we create a situation where another thread may close it underneath resources and leave you with an invalid connection in your hands.

I personally prefer to keep the behaviour consistent, specially because it already involves 2 method calls: `newSessionConnection` and `close`. In this case, if you have a connection, you can used it. Once everyone closes its connections, the http pool will be shutdown.

Thinking about that, perhaps a solution that receives a function rather than creating new connection is more incapsulated and may have less room for mistakes.